### PR TITLE
Second PR to restore reverted commit (#16224)

### PR DESCRIPTION
**Pull request by jjsjann123**
update:
  1. global_reduce check for should_block_y_reduce first.
     This avoids the enabling global_reduce without block_y_reduce. Leading to
     accessing shared memory during global reduce without allocation.
  2. updating block_y_reduce heuristics. Improves perf on tiny tensors
  3. adding test case covering old cases where illegal memory access might occur

  TensorIterator cuda launch configs update (#16224)

    Summary:
    Update launch configs for TensorIterator gpu_reduce_kernel. Enable flexible
    block dimension to improve efficiency for reduction cases with small fast
    dimension.

    Previously TensorIterator launches blocks with fixed 32x16 threads.
    For cases like:

      import torch
      torch.randn(2**20, 4, device='cuda').sum(0)

    The fixed launch config does handle coalesced memory access efficiently.

    Updated launch configure enables flexible block dimension. Combining with
    improved reduction scheme (using flexible vertical / horizontal reduction
    instead of limited warp / block reduction in the old code), it ensures optimal
    memory access pattern even with reduction on dimension with small stride.

    Possible future improvements:
    1. Precise dynamic shared memory allocation.
    2. Using warp shuffle for vertical (block_y) reduction.
    Pull Request resolved: https://github.com/pytorch/pytorch/pull/16224

### CODE REVIEW

**Comment by jjsjann123** [comment]
Benchmark scripts:

```import torch

def bench(size, fn, stride):
   x=torch.ones([size], device='cuda', dtype = torch.float).view(-1, stride)
   torch.cuda.synchronize()
   import time
   nrep = 100
   start = time.time()
   for i in range(nrep):
      fn(x,0)
   torch.cuda.synchronize()
   end = time.time()
   return ((end-start)/nrep)


sum_bws = []
sizes = []
strides = []
for stride in range(1, 4):
  s = 2 ** stride
  #size = 102400
  size = 4096
  while size < 1024000000:
     time = bench(size, torch.sum, s)
     bw = size*4/time*1e-9
     sum_bws.append(bw)
     sizes.append(size)
     strides.append(s)
     size *= 2

for size, stride, sumbw in zip(sizes, strides, sum_bws):
   print(size, stride, sumbw)
```

**Comment by jjsjann123** [comment]
image

Slightly updated heuristics on block y reduction to get better perf

**Comment by jjsjann123** [comment]
Spamming people involved in the second round PR: @soumith @colesbury @umanwizard @fmassa

**Comment by jjsjann123** [comment]
test failure seems to be not machine issue.

**Comment by ezyang** [comment]
It sounds like there is something subtle going on with this patch. Would it be possible to get a comment at the relevant portion of the code, where we made the mistake?

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 1]
```diff
@@ -25,11 +25,21 @@ static inline int64_t div_up(int64_t a, int64_t b) {
   return (a + b - 1) / b;
 }
 
+static inline int last_pow2(int n) {
```
Comment (umanwizard) [inline aten/src/ATen/native/cuda/Reduce.cuh]: Can you give a comment to explain what is going on here?

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 2]
```diff
@@ -331,22 +356,38 @@ struct ReduceOp {
...
 
-  C10_DEVICE arg_t warp_reduce(arg_t value) const {
-    for (int offset = 1; offset < warpSize; offset <<= 1) {
+  C10_DEVICE arg_t block_x_reduce(arg_t value, char* shared_memory) const {
+    int dim_x = blockDim.x;
+    arg_t* shared = (arg_t*)shared_memory;
+    if (dim_x > warpSize) {
+      int address_base = threadIdx.x + threadIdx.y*blockDim.x;
+      shared[address_base] = value;
+      for (int offset = dim_x/2; offset >= warpSize; offset >>= 1) {
```
Comment (umanwizard) [inline aten/src/ATen/native/cuda/Reduce.cuh]: `offset /= 2` seems clearer to me, but it's up to you.

Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Reduce.cuh]: `>>=` is more efficient than `/=` (not that it's a deal breaker tho)
I'm always nervous with division in device code.

**Comment by jjsjann123** [comment]
> It sounds like there is something subtle going on with this patch. Would it be possible to get a comment at the relevant portion of the code, where we made the mistake?

It's really only two lines of change.
I'll post the line number here once I updated it with comments that @umanwizard requested.

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 3]
```diff
@@ -487,38 +528,50 @@ inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t id
...
+    // shuffle instructions and shared memory (if block_width > warpSize).
+    config.input_mult[0] = config.split_input(block_width);
   } else {
     // Otherwise split the output across lanes in a warp.
-    config.output_mult[0] = config.split_output(warp_size);
+    config.output_mult[0] = config.split_output(block_width);
   }
 
-  if (config.values_per_thread() >= warps_per_cta * 16) {
+  if (config.values_per_thread() >= block_height * 16 || config.values_per_thread() >= 256) {
```
Comment (jjsjann123) [inline aten/src/ATen/native/cuda/Reduce.cuh]: This line in the previous PR was 
`if (config.values_per_thread() >= block_height * 16) {`
adapted from the old code `if (config.values_per_thread() >= warps_per_cta * 16) {`

since the old logic applies to a fixed warps_per_cta (16), our adjustable launch config allows `block_height` to be larger than 16. In which case, utilizing block_y_reduction would give significant performance boost.

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 4]
```diff
@@ -487,38 +528,50 @@ inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t id
...
-    config.input_mult[1] = config.split_input(warps_per_cta);
+    config.input_mult[1] = config.split_input(block_height);
   } else {
     // Otherwise, each warp handles a separate output.
-    config.output_mult[1] = config.split_output(warps_per_cta);
+    config.output_mult[1] = config.split_output(block_height);
   }
 
-  if (config.values_per_thread() >= 256 && num_outputs <= 4096) {
+  if (config.input_mult[1] != 0 && config.values_per_thread() >= 256 && num_outputs <= 4096) {
```
Comment (jjsjann123) [inline aten/src/ATen/native/cuda/Reduce.cuh]: This line in the previous PR was not updated. The old code has
`if (config.values_per_thread() >= 256 && num_outputs <= 4096) {`

This is where `illegal memory accessing` is cause. (combining from the last update)
Previously if `config.values_per_thread() >= 256` is true at this phase, it would also enable `block_y_reduction` (aka, `block_reduction`).
With adjustable launch config this is not the case any more.

Hence we would launch a kernel that does global reduction without block y reduction using shared memory. Look at function `shared_memory_size()`. In this case we have 0 shared memory allocated for the kernel, while global reduction accesses shared memory.

**Comment by jjsjann123** [comment]
@ezyang @umanwizard Two lines of change have been pointed out in my comment.

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 5]
```diff
@@ -25,11 +25,22 @@ static inline int64_t div_up(int64_t a, int64_t b) {
   return (a + b - 1) / b;
 }
 
+// returns round down to the closest integer of log2(n)
```
Comment (umanwizard) [inline aten/src/ATen/native/cuda/Reduce.cuh]: This sentence is a bit difficult to understand. I think it means "[floor](https://en.wikipedia.org/wiki/Floor_and_ceiling_functions) of `log2(n)`" -- is that right?

**Review by umanwizard — COMMENTED** [review]
@ezyang this PR seems okay to me, do you have concerns? If not, I will accept and land.

**Comment by ezyang** [comment]
Go for it

File: aten/src/ATen/native/cuda/Reduce.cuh (hunk 5 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Reduce.cuh]: Yep that sounds right.
let me push an update for this

Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Reduce.cuh]: And apologize for my previous confusing & incorrect comment.
